src/main/groovy/com/bugsnag/android/gradle/BugsnagUploadProguardTask.java

```diff
@@ -0,0 +1,130 @@
+        Project project = getProject();
+        File mappingFile = findMappingFile(project);
+        Logger logger = project.getLogger();
+        logger.info("Using mapping file: $mappingFile");
+
+        // If we haven't enabled proguard for this variant, or the proguard
+        // configuration includes -dontobfuscate, the mapping file
+        // will not exist (but we also won't need it).
+        if (mappingFile == null || !mappingFile.exists()) {
+            logger.warn("Mapping file not found: ${mappingFile}");
+            BugsnagPluginExtension bugsnag = project.getExtensions().findByType(BugsnagPluginExtension.class);
+            if (bugsnag.isFailOnUploadError()) {
+                throw new GradleException("Mapping file not found: ${mappingFile}");
+            } else {
+                return;
+            }
+        } else if (mappingFile.length() == 0) { // proguard's -dontobfuscate generates an empty mapping file
+            return;
+        }
+
+        // Read the API key and Build ID etc..
+        BugsnagVariantOutputUtils.readManifestFile(project, variant, variantOutput);
+        logger.info("Attempting to upload mapping file: ${mappingFile}");
+
+        // Construct a basic request
+        Charset charset = Charset.forName("UTF-8");
+        MultipartEntity mpEntity = new MultipartEntity(HttpMultipartMode.BROWSER_COMPATIBLE, null, charset);
+        mpEntity.addPart(partName, new FileBody(mappingFile));
+
+        // Send the request
+        BugsnagMultiPartUploadRequest request = new BugsnagMultiPartUploadRequest();
+        request.applicationId = applicationId;
+        request.variant = variant;
+        request.variantOutput = variantOutput;
+        request.uploadMultipartEntity(mpEntity, getProject());
+    }
+
+    private File findMappingFile(Project project) {
+        if (BugsnagPlugin.hasDexguardPlugin(project) && BugsnagPlugin.hasMultipleOutputs(project)) {
+            File mappingFile = findDexguardMappingFile(project);
+
+            if (mappingFile != null && mappingFile.exists()) {
+                return mappingFile;
+            } else {
+                project.getLogger().warn("Could not find DexGuard mapping file at: $mappingFile -" +
+                    " falling back to AGP mapping file value");
+            }
+        }
+
+        // Use AGP supplied value, preferring the new "getMappingFileProvider" API but falling back
+        // to the old "mappingFile" API if necessary
+        try {
+            FileCollection mappingFileProvider = variant.getMappingFileProvider().getOrNull();
+
+            // We will warn about not finding a mapping file later, so there's no need to warn here
+            if (mappingFileProvider == null || mappingFileProvider.isEmpty()) {
+                return null;
+            }
+            return mappingFileProvider.getSingleFile();
+        } catch (Throwable exc) {
+            return variant.getMappingFile();
+        }
```

Comment:
Should this live in the `AgpCompat` class now?

Reply:
#226 plans on removing the `AgpCompat` class in favour of declaring the compatibility shims where they are used by individual tasks